File: solutions/cpp/yacht/2/yacht.cpp
#include "yacht.h"

#include <map>
#include <algorithm>
#include <stdexcept>

using namespace std;

namespace yacht {

int count_my_number(list<int> dice, int n) {
    int counter = 0;
    for(int d : dice){
        counter += d==n ? 1 : 0;
    }
    return counter;
}
    
int score_by_number(list<int> dice, int n) {
     return count_my_number(dice, n)*n;
}

int score_yacht(list<int> dice){
    return count_my_number(dice, dice.front()) == 5 ? 50 : 0;
}

int score_lt_straight(list<int> dice){
    dice.sort();
    return dice==list<int>{1,2,3,4,5} ? 30 : 0;
}

int score_big_straight(list<int> dice){
    dice.sort();
    return dice==list<int>{2,3,4,5,6} ? 30 : 0;
}

int score_choice(list<int> dice){
    int sum = 0;
    for(int d : dice){
        sum += d;
    }
    return sum;
}

int score_full_house(list<int> dice){
    dice.sort();
    int a = count_my_number(dice, dice.front());
    int b = count_my_number(dice, dice.back());
    return ((a==2 && b==3) || (a==3 && b==2)) ? 
        score_choice(dice)  :  0;        
}

int score_four_of_a_kind(list<int> dice){
     dice.sort();
     int a = count_my_number(dice, dice.front());
     int b = count_my_number(dice, dice.back());
     int sc = 0;
     if(a>=4)
         sc = 4*dice.front();
     else if(b>=4)
         sc = 4*dice.back();
     return sc;
}


int score(std::list<int> dice, std::string cat){ 
    map<string,int> categories = 
        {{"ones", 0}, {"twos", 1}, {"threes", 2}, 
        {"fours", 3},{"fives", 4}, {"sixes", 5}, 
        {"full house", 6}, {"four of a kind", 7},
        {"little straight", 8}, {"big straight", 9}, 
        {"choice", 10}, {"yacht", 11}};
    
    int res = 0;
    auto search = categories.find(cat);
    if (search != categories.end()) {
        switch(search->second) {
            case(0):
            case(1):
            case(2):
            case(3):
            case(4):
            case(5): 
                res = score_by_number(dice, search->second+1);
                break;
            case(6):
                res = score_full_house(dice);
                break;
            case(7): 
                res = score_four_of_a_kind(dice);
                break;
            case(8): 
                res = score_lt_straight(dice);
                break;
            case(9): 
                res = score_big_straight(dice);
                break;
            case(10): 
                res = score_choice(dice);
                break;
            default: 
                res = score_yacht(dice);
        }
    }
    else {
        throw std::invalid_argument("Invalid category.");
    }
     return res;
}

}  // namespace yacht

File: solutions/dart/yacht/2/lib/yacht.dart
import 'package:yacht/categories.dart';
import 'package:collection/collection.dart';

class Yacht {
  List<int> dice;

  Yacht(this.dice){
    dice.sort();
  }

  int count(List<int> ds, int n)=>
    ds.where((d)=>d==n).length;

  int score_fullHouse() {
    int a = count(dice,dice[0]);
    int b = count(dice,dice[4]);
    return ((a==2 && b==3) || (a==3 && b==2)) ? 
      dice.reduce((v,e)=>v+e) : 0; 
  }

  int score_fourOf() {
      if(dice[0]==dice[3] || dice[1] == dice[4])
         return 4*dice[1];
     else return 0;    
  }

   int score_ltlStraight(){
      if(ListEquality().equals(dice,[1,2,3,4,5]))
          return 30;
       else return 0;
   }

   int score_bgStraight(){
      if(ListEquality().equals(dice,[2,3,4,5,6]))
          return 30;
       else return 0;
   }

 int score_yacht(){
    return (dice.toSet().length == 1) ? 50 : 0;} 
  
  int score(Category cat)=>    
    switch(cat){
      Category.ones  => count(dice, 1),
      Category.twos => 2*count(dice, 2),
      Category.threes => 3*count(dice, 3),
      Category.fours => 4*count(dice, 4),
      Category.fives => 5*count(dice, 5),
      Category.sixes => 6*count(dice, 6),
      Category.full_house => score_fullHouse(),
      Category.four_of_a_kind => score_fourOf(),
      Category.little_straight => score_ltlStraight(),
      Category.big_straight => score_bgStraight(),
      Category.choice => dice.reduce((v,e)=>v+e),
      Category.yacht => score_yacht(),
      _ => 0,
    };
  
  
}
